File: backend-app/src/main/java/tom/projects/monkeylog/model/workout/Workout.java
package tom.projects.monkeylog.model.workout;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.OnDelete;
import org.hibernate.annotations.OnDeleteAction;
import tom.projects.monkeylog.model.user.UserOwned;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Stream;

@Entity
@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
public class Workout implements UserOwned {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private Long clonedFromId;

    private String name;

    private String note;

    private Long userId;

    @Enumerated(EnumType.STRING)
    private Type type;

    private LocalDateTime startDate;

    private LocalDateTime endDate;

    @OneToMany(mappedBy = "workout", cascade =  {CascadeType.PERSIST, CascadeType.MERGE })
    @OnDelete(action = OnDeleteAction.CASCADE)
    @OrderColumn(name = "sort_order")
    private List<ExerciseGroup> exerciseGroups = new ArrayList<>();

    @ManyToOne
    @JoinColumn(name="program_week_id")
    private ProgramWeek programWeek;

    public static Workout clone(Workout workout) {
        Workout newWorkout = new Workout();
        newWorkout.setName(workout.getName());
        newWorkout.setUserId(workout.getUserId());
        newWorkout.setNote(workout.getNote());
        newWorkout.setType(workout.getType());
        newWorkout.setProgramWeek(workout.getProgramWeek());
        workout.getExerciseGroups().stream().map(ExerciseGroup::clone).forEach(newWorkout::addExerciseGroup);

        return newWorkout;
    }

    public void addExerciseGroup(ExerciseGroup exerciseGroup) {
        exerciseGroups.add(exerciseGroup);
        exerciseGroup.setWorkout(this);
    }

    public enum Type {
        TEMPLATE,
        ACTIVE,
        COMPLETED
    }
}

File: backend/src/main/kotlin/tom/com/workout/logbook/model/workout/Workout.kt
package tom.com.workout.logbook.model.workout

import jakarta.persistence.*
import org.hibernate.annotations.UuidGenerator
import tom.com.workout.logbook.model.user.UserOwned
import java.time.Instant
import java.util.*

@Entity
class Workout(
    @Id
    @GeneratedValue
    @UuidGenerator
    var id: UUID? = null,
    var clonedFromId: UUID? = null,
    @Column(nullable = false)
    var name: String,
    var note: String? = null,
    override var userId: UUID? = null,
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    var workoutType: tom.com.workout.logbook.model.workout.WorkoutType = tom.com.workout.logbook.model.workout.WorkoutType.TEMPLATE,
    var startDate: Instant? = null,
    var endDate: Instant? = null,
    @OneToMany(mappedBy = "workout", cascade = [CascadeType.ALL], orphanRemoval = true)
    @OrderColumn(name = "sort_order")
    var exerciseGroups: MutableList<tom.com.workout.logbook.model.workout.ExerciseGroup> = ArrayList(),
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "program_week_id")
    var programWeek: tom.com.workout.logbook.model.workout.ProgramWeek? = null,
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "workout_folder_id")
    var workoutFolder: tom.com.workout.logbook.model.workout.WorkoutFolder? = null
) : UserOwned